### ISSUE TITLE
block-scoped-var rule reports duplicate issues in some cases

### ISSUE THREAD

**Issue by ashwinbaliga** [eslint/eslint#17244]
### Environment

Node version: 16.16.0
npm version: 8.11.0
Local ESLint version: 8.41.0 
Global ESLint version: 8.41.0
Operating System: RHEL 7

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```{
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "block-scoped-var": [
            "error"
        ]
    },
    "env": {
        "es2021": true
    }
}

```
</details>

```js
define([], function() {
    if (true) {
      var myvar = 10;
    } else if (true) {
      var myvar = 10;
    } else if (true) {
      var myvar = 10;
    } else if (true) {
      var myvar = 10;
    }
});


```

### What did you expect to happen?

Error for each line should be reported only once

### What actually happened?

Error for each line is reported 3 times

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiZGVmaW5lKFtdLCBmdW5jdGlvbigpIHtcbiAgICBpZiAodHJ1ZSkge1xuICAgICAgdmFyIG15dmFyID0gMTA7XG4gICAgfSBlbHNlIGlmICh0cnVlKSB7XG4gICAgICB2YXIgbXl2YXIgPSAxMDtcbiAgICB9IGVsc2UgaWYgKHRydWUpIHtcbiAgICAgIHZhciBteXZhciA9IDEwO1xuICAgIH0gZWxzZSBpZiAodHJ1ZSkge1xuICAgICAgdmFyIG15dmFyID0gMTA7XG4gICAgfVxufSk7XG5cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiYmxvY2stc2NvcGVkLXZhciI6WyJlcnJvciJdfSwiZW52Ijp7ImVzMjAyMSI6dHJ1ZX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Errors reported:

Error 3:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 3:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 3:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 5:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 5:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 5:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 7:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 7:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 7:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 9:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 9:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

Error 9:11
'myvar' used outside of binding context.  ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

**Comment by Rec0iL99** [comment]
Hi @ashwinbaliga, thanks for the issue. I could reproduce what you described, but I'm unsure if this is intended behaviour or a bug. I'll push it forward for the team to check out.

**Comment by fasttime** [comment]
There are no [unit tests](https://github.com/eslint/eslint/blob/v8.42.0/tests/lib/rules/block-scoped-var.js) with more problems reported than variable usages for this rule, so it seems that the current behavior was not intended/predicted. I think it's a bug.

**Comment by mdjermanovic** [comment]
> Error 3:11 'myvar' used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))
> 
> Error 3:11 'myvar' used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))
> 
> Error 3:11 'myvar' used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

I think this is correct behavior. The first error is for the declaration on line 5, the second is for the declaration on line 7, and the third is for the declaration on line 9. Perhaps we could add the declaration's loc to the error message to make it clear what is reported.

Alternatively, we could update the rule to not report init references, but after that change this rule would produce zero errors in the above example, not one for each line. According to the test cases, reporting init references is intentional:

https://github.com/eslint/eslint/blob/030a82737f51563f9a7b4985fc91b6d8eab54fce/tests/lib/rules/block-scoped-var.js#L163-L176

**Comment by fasttime** [comment]
> I think this is correct behavior. The first error is for the declaration on line 5, the second is for the declaration on line 7, and the third is for the declaration on line 9. Perhaps we could add the declaration's loc to the error message to make it clear what is reported.

Wouldn't it be possible to report one single problem for all declarations? E.g.

    Error 3:11 'myvar' used outside of binding context. Declarations on lines 5, 7 and 9. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

**Comment by mdjermanovic** [comment]
Actually, not reporting init references wouldn't avoid seemingly duplicate errors in all cases. For example, there would still be multiple errors reported on `myvar = 10`:

```js
define([], function() {
    if (true) {
      var myvar;
      myvar = 10; // 3 errors here
    } else if (true) {
      var myvar;
    } else if (true) {
      var myvar;
    } else if (true) {
      var myvar;
    }
});


```

**Comment by mdjermanovic** [comment]
> > I think this is correct behavior. The first error is for the declaration on line 5, the second is for the declaration on line 7, and the third is for the declaration on line 9. Perhaps we could add the declaration's loc to the error message to make it clear what is reported.
> 
> Wouldn't it be possible to report one single problem for all declarations? E.g.:
> 
> Error 3:11 'myvar' used outside of binding context. Declarations on lines 5, 7 and 9. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

That's doable, but seems too complex for addressing this edge case, while it still can be argued that reporting multiple errors is the correct behavior.

**Comment by mdjermanovic** [comment]
I'd suggest we just add the location info to the error message, so that it looks like this (similar to no-shadow messages):

> Error 3:11 'myvar' declared on line 5 column 11 is used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))
> 
> Error 3:11 'myvar' declared on line 7 column 11 is used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))
> 
> Error 3:11 'myvar' declared on line 9 column 11 is used outside of binding context. ([block-scoped-var](https://eslint.org/docs/rules/block-scoped-var))

**Comment by mdjermanovic** [comment]
Looks like there's agreement on https://github.com/eslint/eslint/issues/17244#issuecomment-1580232223. I'm working on this.
